app/models/missing_link_finder.rb

```diff
@@ -7,40 +15,63 @@ class MissingLinkFinder
 
   def initialize(redirection)
     @redirection = redirection
+    @session = Capybara::Session.new(:mechanize)
   end
 
   def run
-    begin
-      response = fetch(redirection.url)
-      if response.is_a?(Net::HTTPSuccess)
-        missing = missing_links(response.body)
-        if missing.empty?
-          { status: :good }
-        else
-          { status: :missing_links, missing: missing }
-        end
+    session.visit(redirection.url)
+    if session.status_code.in?(200..299)
+      if missing_links.empty?
+        {status: :good}
       else
-        { status: :offline }
+        {status: :missing_links, missing: missing_links}
       end
-    rescue SocketError
-      { status: :offline }
-    rescue StandardError => error
-      { status: :error, error: "#{error.class}: #{error.message}" }
+    else
+      {status: :offline}
+    end
+  rescue SocketError
+    {status: :offline}
+  rescue => error
+    if error.message.include?("ERR_NAME_NOT_RESOLVED")
+      {status: :offline}
+    elsif error.message.include?("Net::HTTPNotFound")
+      {status: :not_found}
+    else
+      {status: :error, error: "#{error.class}: #{error.message}"}
     end
   end
 
   private
 
-  attr_reader :redirection
+  attr_reader :redirection, :session
 
-  def missing_links(html)
-    if html
-      missing_links = []
-      missing_links << "next" if !html.match?(next_link)
-      missing_links << "prev" if !html.match?(prev_link)
+  def missing_links
+    found_next = has_next_link?
+    found_previous = has_prev_link?
+    if !found_next || !found_previous
+      session.all("iframe, frame").map { |i| i[:src] }.each do |iframe_src|
+        unless iframe_src.start_with?("http")
+          uri = URI.parse(session.current_url)
+          iframe_src = "#{uri.scheme}://#{uri.host}/#{iframe_src}"
+        end
 
-      missing_links
+        session.visit iframe_src
+        found_next ||= has_next_link?
+        found_previous ||= has_prev_link?
```

Comment:
Just want to talk this one through:

* `found_<link>` is true
  * doesn't change `found_<link>` because the right side of the `||=` condition wouldn't trigger
* `found_<link>` is false because it couldn't find a link earlier
  * it executes `has_<some>_link` because falsey values will execute the right side of the `||=`

Reply:
Yep, exactly. If we've already found a link, we can't **un**find it, so this ensures that if we're searching a page with 3 `<frame>`s and we found a link on the first frame but not the others, it still counts as found rather than it needing to be found on the exact same page every time.